### PR TITLE
feat: add nonce support

### PR DESCRIPTION
**Pull request by DustinJSilk**
fixes https://github.com/BuilderIO/qwik/issues/2663

# What is it?

- [x] Feature / enhancement
- [ ] Bug
- [ ] Docs / tests

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

# Use cases and why

- 1. One use case
- 2. Another use case

# Checklist:

- [ ] My code follows the [developer guidelines of this project](https://github.com/BuilderIO/qwik/blob/main/CONTRIBUTING.md)
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] Added new tests to cover the fix / functionality

### CODE REVIEW

**Comment by manucorporat** [comment]
Interesting, i like how this does not interfere with existing code! love the direction!

**Comment by manucorporat** [comment]
so developers would have to generate their own nonce?
how does this work? it's one for the whole server?

**Comment by DustinJSilk** [comment]
yes they would have to generate it themselves.
They would need to generate it for each new request, and pass it into the render function.

Here is an example i've put together: https://github.com/DustinJSilk/qwik-nonce/blob/main/src/entry.preview.tsx

**Comment by manucorporat** [comment]
could this implemented within qwikcity? as a plugin?

**Comment by manucorporat** [comment]
I understand this solves the problem for you, but we can't change the signature of express middleware so work around to pass a nonce, this needs to work in cloudflare, netlify, vercel...

**Comment by manucorporat** [comment]
I propose to use `requestEv.sharedMap` as a connection point, it could be added a plugin that adds the nonce for every request, and sets the headers and qwik will use it.

How about some well known keys, like:
`@nonce`

**Comment by DustinJSilk** [comment]
> I understand this solves the problem for you, but we can't change the signature of express middleware so work around to pass a nonce, this needs to work in cloudflare, netlify, vercel...

Yep that makes sense!

> ... it could be added a plugin that adds the nonce for every request, and sets the headers and qwik will use it.

Ok, Im happy to work on something like this.
I'll have another crack at it some time today/tomorrow and we can take it from there

Thanks!!

**Comment by DustinJSilk** [comment]
@manucorporat let me know if this is closer to what you had in mind!

[Ive updated the example as well](https://github.com/DustinJSilk/qwik-nonce/blob/main/src/entry.preview.tsx)

**Comment by manucorporat** [comment]
Unfortunately, i didnt mean this. I dont want to add any `nonce` option. instead i wanna rely on `sharedMap` fucntionality to pass a custom nonce to the qwik middleware

**Comment by manucorporat** [comment]
ie, no extra code to handle nonce, it will be implemented by a user plugin to add it

/src/routes/plugin@nonce.ts
```tsx
export const onRequest = (ev) => {
  const nonce = createNonce();
  ev.sharedMap.set('@nonce', nonce):
  ev.headers.set(....)
}
```

File: packages/qwik-city/middleware/request-handler/request-event.ts [hunk 1]
```diff
@@ -31,11 +32,27 @@ export function createRequestEvent(
...
   resolved: (response: any) => void
 ) {
-  const { request, platform, env } = serverRequestEv;
+  const { request, platform, env, contentSecurityPolicy } = serverRequestEv;
 
   const cookie = new Cookie(request.headers.get('cookie'));
-  const headers = new Headers();
+  const headers = new Headers(serverRequestEv.headers);
   const url = new URL(request.url);
+  const nonce = serverRequestEv.nonce ? randomBytes(16).toString('base64') : undefined;
```
Comment (manucorporat) [inline packages/qwik-city/middleware/request-handler/request-event.ts:40]: all this changes should be reverted

File: packages/qwik-city/middleware/request-handler/api.md [hunk 2]
```diff
@@ -151,15 +155,21 @@ export interface ServerRenderOptions extends RenderOptions {
 
 // @alpha
 export interface ServerRequestEvent<T = any> {
+    // (undocumented)
+    contentSecurityPolicy?: ContentSecurityPolicy;
```
Comment (manucorporat) [inline packages/qwik-city/middleware/request-handler/api.md:159]: no changes needed

File: packages/qwik-city/middleware/node/index.ts [hunk 3]
```diff
@@ -31,7 +31,15 @@ export function createQwikCity(opts: QwikCityNodeRequestOptions) {
     next: NodeRequestNextFunction
   ) => {
     try {
-      const serverRequestEv = await fromNodeHttp(getUrl(req), req, res, 'server');
+      const serverRequestEv = await fromNodeHttp(
+        getUrl(req),
+        req,
```
Comment (manucorporat) [inline packages/qwik-city/middleware/node/index.ts:36]: revert all this, no needed

**Comment by DustinJSilk** [comment]
aaaah, i see. Sorry, i didnt fully follow.
Thanks for the time! I'll get this corrected.

**Comment by DustinJSilk** [comment]
Clever! I didnt realise that was actually possible. 
Ive made some updates.

Thanks!

**Comment by manucorporat** [comment]
Amazing feature my friend!!

**Comment by tzdesign** [comment]
@DustinJSilk How did you implement it now? 
It would be nice to document. I created a plugin@nonce.ts and saw actually the @nonce in the map in all the routes, but the script tags lags in the attribute.

Didn't find anything about it in the docs or in discord.
image

Or is this Script-Tag only?  How we are doing with qwiks own script tags or I missing something? Because I would prefer not having unsafe-inline or even unsafe-eval in the headers. 

Am I missing something?

**Comment by DustinJSilk** [comment]
You can see a basic example of the `plugin@nonce.ts` file [here](https://github.com/DustinJSilk/qwik-nonce/blob/main/src/routes/plugin%40nonce.ts)

The sharedMap is a way to share data across all your loaders and endpoints. It's also accessible by the Qwik runtime.
Adding a field `@nonce` to the sharedMap will be automatically picked up by Qwik and added to any internal script tags.

It's then up to you to add that nonce to any scripts in your code base.
The default ones are in the root.tsx and the router-header

Qwik will add the nonce from the sharedMap to the serverData, making it available to your components at render time. You can grab it with this line:

`const nonce = useServerData<string | undefined>("nonce");`

[Here](https://github.com/DustinJSilk/qwik-nonce/blob/main/src/root.tsx) is an example of it being added to the ServiceWorker script. I recommend then passing that into your router-head and adding the nonce to the scripts there

**Comment by DustinJSilk** [comment]
Your browser is likely just hiding the nonce. Try and view the raw page source to see if the nonce exists everywhere.
Or check the console for CSP errors.

**Comment by tzdesign** [comment]
image

You're right! WTF? Chrome

I need to investigate further, CSP is still reporting this vite-error-overlay:
image

Obviously it will not be in production, but I want CSP to work in dev so that all devs know what is happening.


Thank you so much for your support.

**Comment by DustinJSilk** [comment]
Unfortunately I can’t help with that. On my dev environment I don’t use a nonce, I just check it works on a preproduction environment

**Pull request merged**